pam/pam.go

```diff
@@ -73,68 +83,69 @@ func pam_sm_authenticate(pamh *C.pam_handle_t, flags, argc C.int, argv **C.char)
 	p := tea.NewProgram(&appState, opts...)
 	if _, err := p.Run(); err != nil {
 		log.Errorf(context.TODO(), "Cancelled authentication: %v", err)
-		return C.PAM_ABORT
+		return pam.ErrAbort
 	}
 
-	logErrMsg := "unknown"
-	var errCode C.int = C.PAM_SYSTEM_ERR
+	var status = pam.NewTransactionError(pam.ErrSystem, errors.New("unknown exit code"))
 
-	switch exitMsg := appState.exitMsg.(type) {
+	switch exitMsg := appState.exitStatus.(type) {
 	case pamSuccess:
-		brokerIDUsedToAuthenticate = exitMsg.brokerID
-		return C.PAM_SUCCESS
+		if err := mTx.SetData(authenticationBrokerIDKey, exitMsg.brokerID); err != nil {
+			return err
+		}
+		return nil
 	case pamIgnore:
 		// localBrokerID is only set on pamIgnore if the user has chosen local broker.
-		brokerIDUsedToAuthenticate = exitMsg.localBrokerID
-		if exitMsg.String() != "" {
-			log.Debugf(context.TODO(), "Ignoring authd authentication: %s", exitMsg)
-		}
-		logErrMsg = ""
-		errCode = C.PAM_IGNORE
-	case pamAbort:
-		if exitMsg.String() != "" {
-			logErrMsg = fmt.Sprintf("cancelled authentication: %s", exitMsg)
-		}
-		errCode = C.PAM_ABORT
-	case pamAuthError:
-		if exitMsg.String() != "" {
-			logErrMsg = fmt.Sprintf("authentication: %s", exitMsg)
-		}
-		errCode = C.PAM_AUTH_ERR
-	case pamSystemError:
-		if exitMsg.String() != "" {
-			logErrMsg = fmt.Sprintf("system: %s", exitMsg)
+		if err := mTx.SetData(authenticationBrokerIDKey, exitMsg.localBrokerID); err != nil {
+			return err
 		}
-		errCode = C.PAM_SYSTEM_ERR
+		status = exitMsg
+	case pamReturnStatus:
+		status = exitMsg
 	}
 
-	if logErrMsg != "" {
-		fmt.Fprintf(os.Stderr, "Error: %v\n", logErrMsg)
+	if status.Status() != pam.ErrIgnore {
+		if _, err := mTx.StartStringConv(pam.ErrorMsg, status.Error()); err != nil {
+			log.Errorf(context.TODO(), "Failed reporting error to pam: %v", err)
+		}
+	} else if status.Error() != "" {
+		if _, err := mTx.StartStringConv(pam.TextInfo, status.Error()); err != nil {
+			log.Errorf(context.TODO(), "Failed sending info to pam: %v", err)
+		}
 	}
 
-	return errCode
+	return status
 }
 
-// pam_sm_acct_mgmt sets any used brokerID as default for the user.
-//
-//export pam_sm_acct_mgmt
-func pam_sm_acct_mgmt(pamh *C.pam_handle_t, flags, argc C.int, argv **C.char) C.int {
+// AcctMgmt sets any used brokerID as default for the user.
+func (h *pamModule) AcctMgmt(mTx pam.ModuleTransaction, flags pam.Flags,
+	args []string) error {
+	brokerData, err := mTx.GetData(authenticationBrokerIDKey)
+	if err != nil && errors.Is(err, pam.ErrNoModuleData) {
+		return pam.ErrIgnore
+	}
+
+	brokerIDUsedToAuthenticate, ok := brokerData.(string)
 	// Only set the brokerID as default if we stored one after authentication.
-	if brokerIDUsedToAuthenticate == "" {
-		return C.PAM_IGNORE
+	if !ok || brokerIDUsedToAuthenticate == "" {
+		return pam.ErrIgnore
 	}
 
 	// Get current user for broker
-	user := getPAMUser(pamh)
```

Comment:
So, you removed here the helper, which I can understand, but then, you don’t return the same thing in case of error than in the first instance. The first instance did return `if err != nil`:
```
return sendEvent(pam.NewTransactionError(pam.ErrSystem, err))
```
I understand you don’t sendEvent because we are outside of our module, but you don’t wrap it to a transaction error. I guess it’s another good indication that this whole TransactionError is not the way we should go.

Reply:
> I understand you don’t sendEvent because we are outside of our module, but you don’t wrap it to a transaction error. I guess it’s another good indication that this whole TransactionError is not the way we should go.

I don't think it's needed here: the purpose of TransactionError inside `userSelectionModel.Init()` is to send back the error in a parseable format to the module, while here we already have an error that we can just give it back to the module transaction wrapper that will handle it for us. If we do want to return a `pam.ErrSystem` in this case we can still do it though.